### PR TITLE
Add String#byteindex, String#byterindex, and MatchData#byteoffset

### PR DESCRIPTION
**Pull request by shugo**
See https://bugs.ruby-lang.org/issues/13110 for details.

### CODE REVIEW

File: string.c [hunk 1]
```diff
@@ -4170,6 +4252,135 @@ rb_str_rindex_m(int argc, VALUE *argv, VALUE str)
...
+ *
+ *    'foo'.byterindex('o', -1) # => 2
+ *    'foo'.byterindex('o', -2) # => 1
+ *    'foo'.byterindex('o', -3) # => nil
+ *    'foo'.byterindex('o', -4) # => nil
+ *
+ *  Related: String#byteindex.
+ */
+
+    static VALUE
```
Comment (nobu) [inline string.c:4343]:
```suggestion
static VALUE
```

File: re.c [hunk 2]
```diff
@@ -1234,6 +1234,38 @@ match_offset(VALUE match, VALUE n)
...
+{
+    int i = match_backref_number(match, n);
+    struct re_registers *regs = RMATCH_REGS(match);
+
+    match_check(match);
+    backref_number_check(regs, i);
+
+    if (BEG(i) < 0)
+        return rb_assoc_new(Qnil, Qnil);
+    return rb_assoc_new(INT2FIX(BEG(i)), INT2FIX(END(i)));
```
Comment (nurse) [inline re.c:1266]: I think `BEG(i)` is `OnigPosition` which is `ptrdiff_t`. Therefore we need to use `LONG2NUMP.

**Comment by nurse** [comment]
The example implementation for checking byte position is as follows:
```c
diff --git a/string.c b/string.c
index 31c3f11045..4554699861 100644
--- a/string.c
+++ b/string.c
@@ -3979,6 +3979,20 @@ rb_str_index_m(int argc, VALUE *argv, VALUE str)
     return LONG2NUM(pos);
 }

+/* whether given pos is valid character boundary or not
+ * Note that in this function, "character" means a code point
+ * (Unicode scalar value), not a grapheme cluster.
+ */
+bool
+str_check_byte_pos(VALUE str, long pos)
+{
+    const char *s = RSTRING_PTR(str);
+    const char *e = RSTRING_END(str);
+    const char *p = s + pos;
+    const char *pp = rb_enc_left_char_head(s, p, e, rb_enc_get(str));
+    return p == pp;
+}
+
 /*
  *  call-seq:
  *    byteindex(substring, offset = 0) -> integer or nil
@@ -4040,6 +4054,10 @@ rb_str_byteindex_m(int argc, VALUE *argv, VALUE str)
         }
     }

+    if (!str_check_byte_pos(str, pos)) {
+        rb_raise(rb_eArgError, "invalid pos");
+    }
+
     if (RB_TYPE_P(sub, T_REGEXP)) {
         if (pos > RSTRING_LEN(str))
             return Qnil;
```
```shell
./ruby -e'p "あいう".byteindex("う", 4)'
-e:1:in `byteindex': invalid pos (ArgumentError)
	from -e:1:in `<main>'
```

**Comment by shugo** [comment]
> The example implementation for checking byte position is as follows:

Thank you .  I've added the boundary check.

**Review by nurse — APPROVED** [review]
LGTM. Could you add NEWS?

**Pull request merged**